### ISSUE TITLE
Formatting with merge_imports creates invalid Rust

### ISSUE THREAD

**Issue by jonasbb** [rust-lang/rustfmt#3321]
As noted below the before version is already not valid Rust. However, it would be nice if it could be supported because I run into such cases during refactoring or adding new code.
I use `merge_imports` to keep the imports looking clean and sometimes I want to merge different sets of imports. Then it can happen that duplicates arise. This is normally handled well, in that duplicates are merged, just not here.

Before rustfmt:

```rust
use std::fmt::{self, Display};
use std::fmt;

fn main() {
    println!("Hello, world!");
}
```

After rustfmt:

```rust
use std::fmt::{
    self, {self, Display},
};

fn main() {
    println!("Hello, world!");
}
```

rustc error after formatting:

```rustc
error[E0252]: the name `fmt` is defined multiple times
 --> src/main.rs:2:12
  |
2 |     self, {self, Display},
  |     ----   ^^^^ `fmt` reimported here
  |     |
  |     previous import of the module `fmt` here
  |
  = note: `fmt` must be defined only once in the type namespace of this module
```

rustfmt.toml

```toml
merge_imports = true
use_field_init_shorthand = true
```

rustc + rustfmt versions

```sh
$ rustc --version
rustc 1.34.0-nightly (f6fac4225 2019-02-03)
$ rustfmt --version
rustfmt 1.0.1-nightly (be13559 2018-12-10)
```

**Comment by topecongiro** [comment]
Note that since the input is semantically incorrect, rustfmt is not guaranteed to produce the valid Rust code.

**Comment by jonasbb** [comment]
@topecongiro Thanks for pointing out that my example is already invalid code. It was late when I posted this.
I added a small comment to my first message, why I think supporting the merging of duplicate imports is nice, but I agree that it is a low priority.

**Comment by ytmimi** [comment]
`merge_imports` is deprecated. Running this with `imports_granularity=Crate` produces:

**Input**
```rust
use std::fmt::{self, Display};
use std::fmt;

fn main() {
    println!("Hello, world!");
}
```

**Outupt**
```rust
use std::{
    fmt,
    fmt::{self, Display},
};

fn main() {
    println!("Hello, world!");
}
```

Trying to compile either the **Input** or the **Output** with rustc leads to compilation issues. @calebcartwright do you think there is any actions we can take here since `use a;` and `use a::{self};` are different?

**Comment by calebcartwright** [comment]
I'm going to close as there's no action to be taken. The title is misleading, as the only thing that's happening is that rustfmt is not converting _invalid_ rust code to _valid_ rust code by deduping an import, but that is intentional as it's not a change rustfmt can make because doing so is not always semantics preserving and _can_ actually convert valid rust code to invalid rust code.
